File: solo-leveling-ui-service/src/main/java/com/sleepkqq/sololeveling/ui/model/UserData.java
package com.sleepkqq.sololeveling.ui.model;

import java.util.Collection;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.userdetails.UserDetails;

@Getter
@RequiredArgsConstructor
public final class UserData implements UserDetails {

  private final Long id;

  private final String username;

  private final String firstName;

  private final String lastName;

  private final String photoUrl;

  private final Locale locale;

  private final List<UserRole> roles;

  @Override
  public Collection<UserRole> getAuthorities() {
    return roles;
  }

  @Override
  public String getPassword() {
    return null;
  }

  public static UserData fromTgUser(TgUserData tgUser) {
    return new UserData(
        tgUser.id(),
        tgUser.username(),
        tgUser.firstName(),
        tgUser.lastName(),
        tgUser.photoUrl(),
        Optional.of(tgUser.languageCode())
            .filter("ru"::equalsIgnoreCase)
            .map(Locale::forLanguageTag)
            .orElse(Locale.ENGLISH),
        List.of(UserRole.USER)
    );
  }
}

File: solo-leveling-ui-service/src/main/kotlin/com.sleepkqq.sololeveling/ui/model/UserData.kt
package com.sleepkqq.sololeveling.ui.model

import org.springframework.security.core.userdetails.UserDetails
import java.util.Locale


class UserData(
	val id: Long,
	private val username: String?,
	val firstName: String?,
	val lastName: String?,
	val photoUrl: String?,
	val locale: Locale,
	val roles: List<UserRole>
) : UserDetails {

	override fun getAuthorities(): List<UserRole> = roles

	override fun getPassword(): String? = null

	override fun getUsername(): String? = username

	companion object {

		fun fromTgUser(tgUser: TgUserData): UserData {
			return UserData(
				tgUser.id,
				tgUser.username,
				tgUser.firstName,
				tgUser.lastName,
				tgUser.photoUrl,
				tgUser.languageCode
					?.takeIf { "ru".equals(it, ignoreCase = true) }
					.let { Locale.forLanguageTag(it) }
					?: Locale.ENGLISH,
				listOf(UserRole.USER)
			)
		}
	}
}